dataprofiler/data_readers/csv_data.py

```diff
@@ -104,26 +106,145 @@ def _check_and_return_options(self, options):
+        :type preferred: list
+        :param omitted: Delimiters that will be omitted from selection
+        :type omitted: list
+        :param quotechar: Character used for quotes
+        :type quotechar: str
+        :return: delimiter, if none can be found None is returned
+        :rtype: str or None
+        :return: quotechar, if none can be found None is returned
         :rtype: str or None
         """
-        sniffer = csv.Sniffer()
-        sniffer.preferred = [',', '\t', ';']  # removes ' ', ':' from preferred
-        try:
-            dialect = sniffer.sniff(data_as_str)
-        except csv.Error as exc:
-            return None
-        return dialect.delimiter
+
+        # Detect vocabulary (and count)
+        vocab = {}
+        for i in range(len(data_as_str)):
+            c = data_as_str[i]
+            if c not in vocab:
+                vocab[c] = 0
+            vocab[c] += 1
+        if '\n' in vocab: vocab.pop('\n')
+        for char in omitted+[quotechar]:
+            if char in vocab:
+                vocab.pop(char)
+                
+        # Sort vocabulary by count
+        ordered_vocab = []
+        sorted_keys = sorted(vocab, key=vocab.get, reverse=True)
+        for c in sorted_keys:
+            ordered_vocab.append(c)
+
+        if not quotechar:
+            sniffer = csv.Sniffer()
+            sniffer.preferred = preferred
+            try:
+                quotechar = sniffer._guess_quote_and_delimiter(
+                    data_as_str, ordered_vocab[:20])[0]
+            except csv.Error as exc:
+                quotechar = None
+            if not quotechar or len(quotechar) == 0:
+                quotechar = '"'
+                
+        # Evaluate vocab, reviewing rows and columns
+        delimiter = None
+        validated_proposed_delimiters = {}
+        for proposed_delim in ordered_vocab[:20]:
+
+            col_types = {}
+            prior_col_count = None
+            incorrect_delimiter_flag = False
+            cell_type_safe_flag = True
+            row_end_safe_flag = False
+            row_start_safe_flag = False
+            proposed_delim_type = data_utils.detect_cell_type(proposed_delim)
+        
+            # If large dataset, select first 25 rows then random sampling
+            proposed_dataset = data_as_str.split('\n')
+            if len(proposed_dataset) > 25:
+                sample_count = min(int(0.1*(len(proposed_dataset)-25)), 1000)
+                proposed_dataset = proposed_dataset[:25] \
+                    + random.choices(proposed_dataset[:25], k=sample_count)
+
+            # Reverse to start at bottom where likely the most columns
+            # Fewer columns are okay, as long as earlier in file
+            for row_idx in range(len(proposed_dataset)-1, -1, -1):
+                row = proposed_dataset[row_idx]
+                proposed_cells = list(csv.reader(
+                    [row], delimiter=proposed_delim, quotechar=quotechar))[0]
```

Comment:
i wonder if a better solution to quote detection would be to also loop through quotechars like we do delimiters, except limit quotechars to those specified that already exist in the vocab, i.e. `(``, ', ")`